### DISCUSSION TITLE
Building a toolchain or using the latest Clang? How ugly?

### DISCUSSION THREAD

**Discussion by rnikander** [android/ndk#2000] [Q&A]
I'm just getting started with the NDK because I want to port some C++ to Android. I want to use the latest Clang (now 17.0.6), because I'm using C++20 modules and other things, and it drives me nuts not using the latest language and library features. I see Clang 14 in the NDK I have currently. 

I see these pages: 
https://android.googlesource.com/platform/ndk/+/refs/heads/main/docs/Toolchains.md
https://android.googlesource.com/toolchain/llvm_android/+/master/README.md

It sounds rough.

I've built Clang and libc++ from source before, to use on macOS, and tweaked the stdlib namespace so it doesn't conflict with a system one at load time. Can I just do that, and use my own scripts to build an ARM64 shared object to call from JNI, or does this version of Clang in the NDK have more modifications to produce Android compatible binaries?

**Comment by rnikander** [comment]
Or maybe there's a way to use Visual Studio for Android dev and use their C++ compiler? I'm not attached to Clang. I just want to use the latest C++.

**Comment by DanAlbert** [comment]
> I see Clang 14 in the NDK I have currently.

So upgrade your NDK? https://github.com/android/ndk/wiki/Changelog-r26 is 17ish.

> It sounds rough.

That's what it takes to build a full multi target toolchain. It's easier to build just the macOS toolchain because you've only done half the work. The compiler isn't the only part of the NDK. You need a sysroot too. That's _why_ there's an NDK and this page isn't just a link to the clang downloads.

**Reply by rnikander** [reply]
Doh, I thought I had the latest. 🤦‍♂️It seems the gradle build insists on downloading and using an older one, but I should be able to figure out how to stop that. The SDK Tools tab in Settings shows "CMake 3.22" as the highest available. I think I need 3.28 if I want to use C++ modules, maybe I can install that manually and point gradle at it.

**Reply by DanAlbert** [reply]
I strongly suggest reading the docs: https://developer.android.com/studio/projects/install-ndk
